### PR TITLE
The complete comment system revamp

### PR DESCRIPTION
**Pull request by dwitterer**
(Don't merge yet!)

I was initially going to create just a separate branch for the tests for the new hashtag system, but I ended up having all the new changes in this branch. Anyway, this pull request covers all the new functionalities and improvements for the comment system on Dwitter:

* The "anchor annihilator", as I affectionately call it, will now remove all links from code blocks. So stuff like colour codes and URLs will not be transformed into <a> tags - this makes hashtags more accurate, and allows users to use code blocks as a mechanism to escape mentions (u/, d/) and links
* Long, dwitter-specific, URLs will now be "autocropped" into their shorter form upon comment submission - for example, `dwitter.net/d/123` will automatically turn into `d/123`
* Hashtag requirements updated to allow tags starting with digits - tags must now be preceded by a space (or start of string), and must include at least one letter; old restrictions still apply for the characters allowed in said hashtags
* New tests were added, per request, to make sure that all the new features work correctly (I know you don't care)

NOTE: The first two bulletpoints seem to require a refresh to work, after you post a comment using them... I'm not sure if this will be the case in production, but it's not a major issue anyway. Just thought I'd let you know, so that you would know. They should show up normally for other users viewing comments.

Overall, just minor things really, but they fix many of the bugs and annoying technical issues with the current comment system, and greatly increase the accuracy of the hashtag system.

### CODE REVIEW

**Comment by dwitterer** [comment]
Things I still need to do with this PR:
* Fix linting errors
* See if test pass as planned
* Get lionleaf's go ahead

Things to fix with the comment system in future PRs:
* The refresh thing, if applicable
* Hashtags followed by a dot (and possibly some other characters - I might have an idea how to fix this, but I'll hold off with it for now)
* Anything else? I can't think of any more issues with the comment system :P

File: dwitter/tests/templatetags/test_insert_magic_links.py [hunk 1]
```diff
@@ -20,7 +21,7 @@ def test_insert_magic_links_replaces_user_with_valid_characters(self):
     def test_insert_magic_links_bypasses_user_with_invalid_characters(self):
         self.assertEqual(
             'u/a1$',
-            'u/a1$'
+            insert_magic_links('u/a1$')
```
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: Yeah, but now a test fails... So I have to either switch it back, or introduce a new regex test to avoid inserting anchors into partially correct usernames. I have came up with such a regex, but users would have to insert a space after the username. So I thought about allowing dots and commas after the username as well, but I don't know if it's all too good to deal with this issue like that. I said this on discord as well, but noone replied :/

File: dwitter/templatetags/insert_magic_links.py [hunk 2]
```diff
@@ -4,6 +4,20 @@
 register = template.Library()
 
 
+# find all long d/ and u/ links in http:// form through regex, and
+# use replace() to singularly shorten each match into the d/ or u/ form
+def autocrop_urls(m):
```
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/templatetags/insert_magic_links.py]: Thanks :)

File: dwitter/templatetags/insert_code_blocks.py [hunk 3]
```diff
@@ -9,6 +9,11 @@ def to_code_block(m):
     code = m.group('code')
     code = re.sub(r'\\`', '`', code)
 
+    # remove anchors from code blocks
```
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/templatetags/insert_code_blocks.py]: No, this is to remove all links from code blocks. The ordering change was put in place to make this work. Not vice versa :P

File: dwitter/tests/templatetags/test_insert_magic_links.py [hunk 4]
```diff
@@ -71,6 +72,76 @@ def test_insert_magic_links_replaces_user_inside_parenthases(self):
...
 
+    # autocrop (http://dwitter.net/d/1 -> d/1)
+
+    def test_insert_magic_autocrops_urls_d(self):
+        self.assertEqual(
+            '<a href="/d/123">d/123</a>',
+            insert_magic_links('dwitter.net/d/123')
+        )
+
+    def test_insert_magic_autocrops_urls_d_http(self):
```
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: I do test both `http` and `https`... As well as with and without `www.`. I don't know about the same test though, I might need to have a look later. Not my main priority right now though.

File: dwitter/tests/templatetags/test_insert_magic_links.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lionleaf) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: I'm ok with `u/a1$` becoming `<a href='/u/a1'>u/a1</a>$` which I believe is how it works today?

You can change this test to `insert_magic_links_does_not_linkify_invalid_username_characters` or something similar and change it to the way it works now.

File: dwitter/tests/templatetags/test_insert_magic_links.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lionleaf) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: Where do you test with `https`? Am I just not seeing it?

File: dwitter/templatetags/insert_code_blocks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lionleaf) [reply dwitter/templatetags/insert_code_blocks.py]: Got it

**Comment by dwitterer** [comment]
Thanks for the info. I will certainly have a look at the models.py. The refreshing thing woud be qood to fix, but I think that I'll move that onto another PR. This is already quite a large one, and minor annoyances like this are not a major priority right now. I will also have a look at the code thing you sent, but your formatting kinda messed up there, so not sure exactly what that is supposed to say. As for the number-only hashtags, the requirement for a letter kinda put itself in place, and I just kept it there, as it seems to match the way twitter hashtags work, and y'know, twitter knows their stuff when it comes to hashtags :wink: Also, number-only hashtags probably wouldn't be very meaningful anyway. Also also, wht if I make a comment saying "My dweet is #1"? I don't want the "#1" to be a hashtag!

Anyways, I have quite a lot of school work to do for the next few days, but hopefully I can have a look (and maybe *finally* get this PR ready) next friday/weekend :)

Thanks again for the stuff!

File: dwitter/tests/templatetags/test_insert_magic_links.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: Yeah, but it's kind of annoying, and doesn't feel right. If a username contains invalid characters, why match it? It's invalid!

I dunno... maybe I can find a better way of handling this.

File: dwitter/tests/templatetags/test_insert_magic_links.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: Look at the whole file, I have set up specific tests for all the diferent types of URLs.

File: dwitter/tests/templatetags/test_insert_magic_links.py [hunk 5]
```diff
@@ -133,46 +204,54 @@ def test_insert_magic_links_replaces_dweet_in_parenthases(self):
...
         self.assertEqual(
-            'Dwitter is just <a href="/h/amaze">#amaze</a>-balls, right?',
+            'Dwitter is just #amaze-balls, right?',
             insert_magic_links('Dwitter is just #amaze-balls, right?')
         )
 
-    def test_insert_magic_hashtag_not_start_with_digit(self):
+    def test_insert_magic_ignores_hashtag_no_letters(self):
+        self.assertEqual(
+            'Dwitter is so #1337',
```
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: What are you referring to here? There are quite a few changes in this commit...

File: dwitter/tests/templatetags/test_insert_magic_links.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: Oh crap, you're right! I've made tests for http, www, and a combination of the two but no https :O... Luckily, that's a pretty easy fix :P... The regex handles it already, so it's just 4 more tests (https with and without www. for d/ and u/ links).

File: dwitter/tests/templatetags/test_insert_magic_links.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: Well, with hashtags it's whitespace or nothing, and that's before the hashtag :/... And then with endings there's the whole ordeal of what to define as punctuation, and what's just an invalid character. But as I said, I need to have a think about this.

File: dwitter/tests/templatetags/test_insert_magic_links.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwitterer) [reply dwitter/tests/templatetags/test_insert_magic_links.py]: Fixed. Major improvements made to the tests file in general, but I also made the tests default to https://, with a specific section of tests for http:// as well.

**Comment by dwitterer** [comment]
Okay, so I fixed the formatting in that code comment of yours, and pasted it into my localhost version of dwitter. The only thing that I could find that could possibly fail is the double backticks as in \`whatever \`a\`\`, and we already test for that, so I'll just ignore that for now, unless you can clarify what issue you have found.

So all I need to do now, is solve the usernames thing....

That was easy :D

**Comment by dwitterer** [comment]
Okay, so I fixed all of the stuff you mentioned above, and put tests in place for the new punctuation stuff. As of right now, all d/ links, u/ links, and hashtags have to be followed by a space, end-of-string, comma, colon, semi-colon, question mark, or exclamation point. Parentheses and quotes will be added in soon to fix 2-3 more tests....

However, **I kinda need your help**, as half of my tests are now failing due to 'unbalanaced parentheses' in the same file, at the same place, when I haven't even touched that file... :thinking: 

After we fix that, we're pretty much ready to merge :D

**Comment by stianjensen** [comment]
The error is on this line:
https://github.com/lionleaf/dwitter/pull/424/files#diff-f552aa4b1485a67b4f2ebbdd750855c0R138

File: dwitter/models.py [hunk 6]
```diff
@@ -135,8 +135,7 @@ def __str__(self):
 # Should be idempotent.
 @receiver(post_save, sender=Comment, dispatch_uid="add_hashtags_from_comment")
 def add_hashtags(sender, instance, **kwargs):
-    hash_pattern = re.compile(r'#(?P<hashtag>[_a-zA-Z][_a-zA-Z\d]*)')
-    for hashtag in re.findall(hash_pattern, instance.text):
+    for hashtag in re.findall(r'(?<!\S)#([_a-zA-Z\d]+)\b)', instance.text):
```
Comment (stianjensen) [inline dwitter/models.py]: Did you intend to have a `)` both before and after the final word boundary here? Maybe the last one should go away?

```suggestion
    for hashtag in re.findall(r'(?<!\S)#([_a-zA-Z\d]+)\b', instance.text):
```

Comment (dwitterer) [reply dwitter/models.py]: Yeah... `')` not `)'`... Thanks :)

**Comment by lionleaf** [comment]
Looks like I messed up github markdown in an earlier comment. Let me try again; what happens with a comment like this (today #FFFFFF would become a clickable hashtag):

```
This is a comment with some code: ` color:#FFFFFF;` 
```

Does it look ok? And I guess it might be silently added to hashtags, so even if #FFFFFF doesn't become clickable the dweet can now be found at h/FFFFFF  ?

**Comment by lionleaf** [comment]
Should fix #334
